aidants_connect_web/signals.py

```diff
@@ -14,4 +15,8 @@ def on_mandat_change(sender, instance, created, **kwargs):
     if created:
         Journal.objects.mandat_creation(instance)
     else:
-        Journal.objects.mandat_update(instance)
+        update_fields = kwargs.get("update_fields")
```

Comment:
J'ai peur que cette implémentation soit un peu flaky
1. Est-ce qu'il peut il y avoir created==False ET kwargs.get("update_fields")==False ?
--> si non, en ligne 19, on pourrait avoir 
```py
if "last_used_date" in update_fields:
       Journal.objects.mandat_use(instance)
```
2. J'ai peur dans cette implémentation qu'on rate des cas où d'autres champs seraient mis à jour, ou last_used_date ET un autre champs sont mis à jour, on ne journalier pas la mise à jour de l'autre champs.

Reply:
Je ne suis pas sûr de bien avoir saisi 100% du contexte ici, donc peut-être que ce commentaire sera complètement hors-sujet, mais j'ai l'impression qu'on est dans une situation où on cherche à _effectuer une action si un champ spécifique d'un modèle a été modifié_, alors je me permets de vous signaler que [`django-model-utils`](https://django-model-utils.readthedocs.io/en/latest/) propose, entre autres choses bien pratiques, un [`FieldTracker`](https://django-model-utils.readthedocs.io/en/latest/utilities.html#field-tracker) qui offre ce genre de possibilités 😄

Reply:
en fait il y a plusieurs choses qui entrent en jeu ici, et je ne vois pas de solution parfaite (la doc sur `post_save` est [ici](https://docs.djangoproject.com/en/3.0/ref/signals/#post-save)):
- la variable `created` est un boolean, soit True si on créé l'objet, soit False si on modifie l'object
- `update_fields` peut être None (tout dépend si il est passé comme paramètre dans le .save()), c'est pour ça que je dois faire un check au début de la ligne 19 (sinon ca plante)
- en creusant un peu, en fait lorsqu'on fait des modifs dans l'admin, par défaut le paramètre `update_fields` n'est pas envoyé dans le `.save()`, donc on saura que l'objet a été créé/modifié, mais pas quels champs ont été modifiés
- d'où en effet l'idée proposée par Mathieu d'avoir un tracker dédié. ou d'overwrite le post_save de l'admin (comme [ici](https://stackoverflow.com/questions/54578488/django-signals-kwargsupdate-fields-is-always-none-on-model-update-via-djan))
- on a aussi la possibilité de faire tout ca dans le `pre_save` (doc [ici](https://docs.djangoproject.com/en/3.0/ref/signals/#pre-save), comparer le nouvel objet avec l'objet précedent, et appeler le/les journal correspondants. exemple [ici](https://stackoverflow.com/a/54653160/4293684). avec l'avantage d'attraper à la fois modifs dans les views que dans l'admin + mettre toute la logique dans signals. mais léger risque que ca devienne usine a gaz. après on peut faire une suite de if (au lieu des if else)

Reply:
Personnellement, [pour ce genre de raisons](https://lincolnloop.com/blog/django-anti-patterns-signals/), je ne suis en règle générale pas super fan de l'utilisation des signaux 😅

Reply:
merci pour l'article ! je vais essayer de faire ca en overwrittant le save :)